content/en/blog/_posts/2022-04-19-csi-nodeexpandsecret.md

```diff
@@ -0,0 +1,190 @@
+that made use of [StorageClass Secrets](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html),
+but you didn't have any mechanism to specify the credentials that would be used for
+operations that take place when the storage was mounted onto a node and when
+when the volume has to be expanded at node side.
+Even-though Kubernetes CSI have implemented similar mechanism for Controller
+side operations, ie secretRef field available in the csi PV source and making
+use of it while controllerExpand request has been sent to the CSI driver, similar
+field was missing in the nodeExpandVolume request.  This was causing some trouble
+for different CSI drivers to fulfill below scenarios while it perform NodeExpansion
+operation.
+
+- At times, the CSI driver needs to check the actual size of the backend block storage (or image)
+  before proceeding with a node-level filesystem resize operation. This avoids false positive returns
+  from the backend storage cluster during filesystem resizes.
+
+- When a PersistentVolume represents encrypted block storage (for example using LUKS)
+  you need to provide a passphrase in order to resize the device, and also to make it possible
+  to grow the filesystem on that device.
+
+- For various validations at time of node expansion the CSI driver has to be connected
+  to the backend storage cluster, if the secretRef is part of the
+  nodeExpandVolume request the CSI driver can make use of the same and connect to
+  the storage cluster  to perform the cluster operations.
+
+## How does it work?
+
+To enable this functionality from this version of Kubernetes, Kubernetes Storage SIG have introduced
+a new feature gate called `CSINodeExpandSecret`. Once the feature gate is enabled
+in the cluster, NodeExpandVolume requests can include a `secretRef` field. The NodeExpandVolume request
+is part of CSI; for example, in a request which has been sent from the Kubernetes
+control plane to the CSI driver.
+The admin can specify these secrets as an opaque parameter in storage class as
+he/she specifies other CSI secrets today.  The secrets mention in the storage class
+has to carry the key name as shown below
+
+```
+csi.storage.k8s.io/node-expand-secret-name: test-secret
+csi.storage.k8s.io/node-expand-secret-namespace: default`
+```
+
+If feature gates are enabled and storage class carries the above secret configuration,
+the subject CSI provisioner receives the credentials from the Secret as part of the NodeExpansion request.
+
+CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
+field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.
+
+
+
+## Trying it out
+
+
+Step 1: enable `CSINodeExpandSecret` feature gate (Please refer to
+[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/))
+
+Step 2: Secret and storage class creation
```

Comment:
I'd write
1. Enable the `CSINodeExpandSecret` feature gate (please refer to
[Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)).

1. Create a Secret, and then a StorageClass that uses that Secret
```
(Markdown does automatic numbering)